### ISSUE TITLE
Design: Support Azure Repos Policies

### ISSUE THREAD

**Issue by nmiodice** [microsoft/terraform-provider-azuredevops#83]
As a developer managing an instance of AzDO, I want to enable branch protection policies for the repositories hosted in AzDO so that I can ensure that builds pass prior to being merged into a branch that I want to protect.

This will be quite a bit of work so this task requires a deep dive to understand the APIs that we need to invoke and then it needs to be broken into individual design or implementation stories

# Relevant APIs
APIs backing these operations are here: https://docs.microsoft.com/en-us/rest/api/azure/devops/policy/?view=azure-devops-rest-5.0

# Related Issues
This is blocked on support for Azure Repos: https://github.com/microsoft/terraform-provider-azuredevops/issues/73

Support:
 - Protect this branch w/ Require a minimum number of reviewers
 - Build Validation

# Acceptance Criteria
 - [x] Document resource schema
 - [x] Document validation rules for how to validate configurations 
 - [x] How will we decide what changes can be updated vs require a delete/create to occur?
 - [x] Discuss key implementation details that are relevant
 - [x] Design should live in this issue (the comments) and be reviewed by the core contributor team. It should cover the items above as well as any other design considerations that impact the implementation
 - [ ] Implementation story breakdown into backlog, **after the design review has been completed**

**Comment by nmiodice** [comment]
Based on this feature, perhaps PR policies can be de-prioritized for now https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema#pr-trigger

**Comment by pascalberger** [comment]
@nmiodice I don't think this is related. The YAML schema allows to define pull request triggers for builds, but there's not possibility to define branch policies through it.

It would be nice to have the possibility to manage branch policies by code. Especially if it supports some kind of inheritance (e.g. define a branch policy on project level which would be set for all repos in this project). The current approach with configuration through UI doesn't really work once a projects, repos and branches need to be managed. We have written our custom tool for it: https://github.com/bbtsoftware/AzureDevOpsPolicyConfigurator. To have a terraform provider with the same capabilities would be great.

**Comment by nmiodice** [comment]
@Pascalberger, I'm inclined to agree based on the documentation. Thanks for pointing me towards it :)

Can you point out the code for implementing branch policies in https://github.com/bbtsoftware/AzureDevOpsPolicyConfigurator?

**Comment by pascalberger** [comment]
@nmiodice In which part are you interested? The actual update of policies happens [here](https://github.com/bbtsoftware/AzureDevOpsPolicyConfigurator/blob/develop/src/AzureDevOpsPolicyConfigurator/Logic/PolicyExecuter.cs#L73). It utilizes Azure DevOps .NET client libraries to call the REST API. The resolving of the input configuration JSON file happens in [this method](https://github.com/bbtsoftware/AzureDevOpsPolicyConfigurator/blob/develop/src/AzureDevOpsPolicyConfigurator/Logic/PolicyExecuterBase.cs#L50-L102) (and the helpers which are called from there).

**Comment by nmiodice** [comment]
Here are some **WIP** notes to come back to later...
```bash
AZDO_PAT=<...>
POLICY_NAME=

# Lookup Group
POLICY_ID=$(
    curl -s -u ":$AZDO_PAT" https://dev.azure.com/terraform-azdo/terraform-provider-azuredevops/_apis/policy/types | \
    jq -r ".value | .[] | select(.displayName == \"$POLICY_NAME\") | .id"
)
echo "Found Policy ID: $POLICY_ID"
```

**Comment by nmiodice** [comment]
**WIP:** Possible resource configurations

```hcl
resource "azuredevops_policy_build" "p1" {
    // bool | optional | ForceNew := false
    // OneOf: true (default), false
    enabled = ...

    // bool | optional | ForceNew := false
    // OneOf: true (default), false
    blocking = ...

    settings {
        // string | required | ForceNew := false
        build_definition_id = ...

        // bool | optional | ForceNew := false
        // OneOf: true (default), false
        queue_on_source_update = ...
        
        // int | optional | ForceNew := false
        // default 0
        valid_duration = ...
    }

    scope {
        // string | required | ForceNew := false
        repository_id = ...

        // string | required | ForceNew := false
        repository_ref = ...

        // string | optional | ForceNew := false
        // OneOf: "Exact" (default), "Prefix"
        match_type = ...
    }
}

resource "azuredevops_policy_min_reviewers" "p2" {
    // bool | optional | ForceNew := false
    // OneOf: true (default), false
    enabled = ...

    // bool | optional | ForceNew := false
    // OneOf: true (default), false
    blocking = ...

    settings {
        // int | required | ForceNew := false
        reviewer_count = ...

        // bool | optional | ForceNew := false
        // OneOf: true, false (default)
        submitter_can_vote = ...
    }

    scope {
        // string | required | ForceNew := false
        repository_id = ...

        // string | required | ForceNew := false
        repository_ref = ...

        // string | optional | ForceNew := false
        // OneOf: "Exact" (default), "Prefix"
        match_type = ...
    }
}
```

**Comment by pascalberger** [comment]
@nmiodice Do you plan to also support custom status (status set through the REST API on a pull request)?

**Comment by nmiodice** [comment]
@pascalberger in the long term I can see this being supported, but for the initial implementation I'll be focusing on build validation & minimum # of reviewers.

My hope is to build this in a way that makes supporting other types of build policies very simple... more to come when I have time to consolidate my notes.

**Comment by nmiodice** [comment]
@pascalberger see the updated design -- I appreciate feedback!

**Comment by nmiodice** [comment]
@KeithJRome  I have updated based on our design review

**Comment by nmiodice** [comment]
Note: The composition approach in this branch is more extensible: https://github.com/microsoft/terraform-provider-azuredevops/compare/thephen/issue-160-github-sep

The design should reflect that -- exposing base flatteners/expanders that can (should be) be used by the subtypes, but not forcing a composition pattern on them. Need to update the design to reflect this before closing

**Comment by nmiodice** [comment]
Implementation stories:
 - https://github.com/microsoft/terraform-provider-azuredevops/issues/182
 - https://github.com/microsoft/terraform-provider-azuredevops/issues/184
 - https://github.com/microsoft/terraform-provider-azuredevops/issues/183
